cpp/src/parquet/arrow/reader.h

```diff
@@ -249,6 +249,13 @@ class PARQUET_EXPORT FileReader {
 
   virtual ::arrow::Status ReadRowGroup(int i, std::shared_ptr<::arrow::Table>* out) = 0;
 
+  virtual ::arrow::Status WillNeedRowGroups(const std::vector<int>& row_groups,
+                                            const std::vector<int>& column_indices) = 0;
```

Comment:
It can't be expressed in this API. This method is translated into call of arrow::io::RandomAccessFile::WillNeed()

No-op is default and valid implementation of WillNeed. It means that no preload/prefetch is provided in this RAF implementation. All work will be done when ReadAt or ReadAsync is called.

Current Arrow API expect tight coupling between FileReader, ParquetFileReader and intermediate Cache. It is not possible to provide true async decoupling w/o significant API changes (it was discussed somewhere).

For my technique to work, one should provide special implementation of arrow::io::RandomAccessFile which will receive WillNeed, download the data and signals it in some "hidden" way. Not perfect, but possible to reach what I needed w/o API changes and any other side effects.

 I think I'll be able to provide you link ro real use case tomorrow.

Reply:
https://github.com/apache/arrow/pull/14723 adds a filesystem method for "read many".  I would like to see this method support plugging and splitting in the same way that `ReadRangeCache` does today (then, `ReadRangeCache` will only be needed if you need true "caching").  Then I think we can use that instead of the `ReadRangeCache`.

This will allow local filesystems to rely on the OS for plugging & splitting and will allow remote filesystems like S3 to adapt the algorithm to their needs.  It's also async and returns a future reliably so you can then return a future from this method (I agree that would be desired).

Reply:
+1 for @westonpace's suggestion.

In addition, what if `WillNeedRowGroups` (w/ or w/o same inputs) has been called more than once? Maintaining the state is rather tricky according to my experience. If the new function only issues I/O hints to the `RandomAccessFile`, probably it is much easier to reason about the behavior directly from `RandomAccessFile`.